Source/Events.Store.MongoDB/Processing/Streams/StreamProcessorStates.cs

```diff
@@ -27,60 +29,125 @@ public class StreamProcessorStates : IStreamProcessorStateRepository
-                //     .FirstOrDefaultAsync(cancellationToken)
-                //     .ConfigureAwait(false);
+                if (id is SubscriptionId subscriptionId)
+                {
+                    if (baseStreamProcessorState is StreamProcessorState streamProcessorState)
+                    {
+                        var replacementState = new SubscriptionState(
+                            subscriptionId.ConsumerTenantId,
+                            subscriptionId.ProducerMicroserviceId,
+                            subscriptionId.ProducerTenantId,
+                            subscriptionId.ScopeId,
+                            subscriptionId.StreamId,
+                            subscriptionId.PartitionId,
+                            streamProcessorState.Position,
+                            streamProcessorState.RetryTime,
+                            streamProcessorState.FailureReason,
+                            streamProcessorState.ProcessingAttempts);
+                        var states = await _connection.GetSubscriptionStateCollection(replacementState.ScopeId, cancellationToken).ConfigureAwait(false);
+                        var persistedState = await states.ReplaceOneAsync(
+                            CreateFilter(subscriptionId),
+                            replacementState,
+                            new ReplaceOptions { IsUpsert = true })
+                            .ConfigureAwait(false);
+                    }
+                }
+                else if (baseStreamProcessorState is Partitioned.StreamProcessorState partitionedStreamProcessorState)
+                {
+                    var streamProcessorId = id as StreamProcessorId;
+                    var states = await _connection.GetStreamProcessorStateCollection(streamProcessorId.ScopeId, cancellationToken).ConfigureAwait(false);
+                    var state = await states.ReplaceOneAsync(
+                        CreateFilter(streamProcessorId),
+                        partitionedStreamProcessorState,
+                        new ReplaceOptions { IsUpsert = true })
+                        .ConfigureAwait(false);
+                }
+                else if (baseStreamProcessorState is StreamProcessorState streamProcessorState)
+                {
+                    var streamProcessorId = id as StreamProcessorId;
+                    var states = await _connection.GetStreamProcessorStateCollection(streamProcessorId.ScopeId, cancellationToken).ConfigureAwait(false);
+                    var state = await states.ReplaceOneAsync(
+                        CreateFilter(streamProcessorId),
+                        streamProcessorState,
+                        new ReplaceOptions { IsUpsert = true })
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    throw new CannotPersistStreamProcessorState(id, baseStreamProcessorState);
```

Comment:
It may make sense to throw a more explicit exception. Like 'StreamProcessorStateOfUnsupportedType' or something like that